File: 01_CODEBASES/localbrain-electron/renderer/lib/oklch.ts
// OKLCH Color System - TypeScript Port
// Perceptually uniform color space for LocalBrain

export interface OKLCH {
  L: number;
  C: number;
  h: number;
}

/**
 * Convert OKLCH to RGB hex string
 * @param L Lightness (0-1)
 * @param C Chroma/saturation (0-0.4)
 * @param hDeg Hue in degrees (0-360)
 */
export function oklch(L: number, C: number, hDeg: number): string {
  const h = (hDeg * Math.PI) / 180;
  const a = C * Math.cos(h);
  const b = C * Math.sin(h);

  // OKLab to Linear RGB
  const l_ = L + 0.3963377774 * a + 0.2158037573 * b;
  const m_ = L - 0.1055613458 * a - 0.0638541728 * b;
  const s_ = L - 0.0894841775 * a - 1.291485548 * b;

  const l = l_ * l_ * l_;
  const m = m_ * m_ * m_;
  const s = s_ * s_ * s_;

  // Linear RGB to sRGB
  const toSRGB = (x: number): number => {
    const y = Math.max(0, Math.min(1, x));
    return y <= 0.0031308 ? 12.92 * y : 1.055 * Math.pow(y, 1 / 2.4) - 0.055;
  };

  const r = toSRGB(+4.0767416621 * l - 3.3077115913 * m + 0.2309699292 * s);
  const g = toSRGB(-1.2684380046 * l + 2.6097574011 * m - 0.3413193965 * s);
  const b_ = toSRGB(-0.0041960863 * l - 0.7034186147 * m + 1.707614701 * s);

  // Convert to hex
  const toHex = (n: number) =>
    Math.round(n * 255)
      .toString(16)
      .padStart(2, "0");
  return `#${toHex(r)}${toHex(g)}${toHex(b_)}`;
}

// ===== APCA CONTRAST ALGORITHM =====
// Accessible Perceptual Contrast Algorithm (WCAG 3.0)
// More accurate than WCAG 2.x contrast ratios

/**
 * Calculate APCA contrast between text and background
 * @param textL Lightness of text (0-1)
 * @param bgL Lightness of background (0-1)
 * @returns Lc contrast value (0-106 for normal polarity)
 */
export function calculateAPCA(textL: number, bgL: number): number {
  // Constants from APCA 0.0.98G specification
  const normBG = 0.56;
  const normTXT = 0.57;
  const revTXT = 0.62;
  const revBG = 0.65;
  const blkThrs = 0.022;
  const blkClmp = 1.414;
  const scaleBoW = 1.14;
  const scaleWoB = 1.14;
  const loConThresh = 0.1;
  const loConFactor = 27.7;
  const loConOffset = 0.027;
  const deltaYmin = 0.0005;

  // Soft clamp for Y values
  const clampY = (Y: number): number => {
    return Y >= blkThrs ? Y : Y + Math.pow(blkThrs - Y, blkClmp);
  };

  // Apply soft clamp
  const Ytxt = clampY(textL);
  const Ybg = clampY(bgL);

  // Determine polarity (light text on dark vs dark text on light)
  let SAPC = 0;

  if (Math.abs(Ybg - Ytxt) < deltaYmin) {
    return 0; // Too similar
  }

  if (Ybg > Ytxt) {
    // Dark text on light background (normal polarity)
    SAPC = (Math.pow(Ybg, normBG) - Math.pow(Ytxt, normTXT)) * scaleBoW;
  } else {
    // Light text on dark background (reverse polarity)
    SAPC = (Math.pow(Ybg, revBG) - Math.pow(Ytxt, revTXT)) * scaleWoB;
  }

  // Low contrast threshold
  if (Math.abs(SAPC) < loConThresh) {
    return 0;
  }

  // Scale and offset
  if (SAPC > 0) {
    SAPC = SAPC - loConOffset;
  } else {
    SAPC = SAPC + loConOffset;
  }

  return SAPC * 100;
}

/**
 * Check if APCA contrast meets minimum requirements
 * @param Lc APCA contrast value
 * @param fontSize Font size in px
 * @param fontWeight Font weight (400 = normal, 700 = bold)
 * @returns true if contrast is sufficient
 */
export function meetsAPCAStandard(
  Lc: number,
  fontSize: number = 16,
  fontWeight: number = 400
): boolean {
  const absLc = Math.abs(Lc);

  // APCA guidelines for minimum contrast
  // Body text (16px, normal): Lc 60+
  // Large text (24px, normal): Lc 45+
  // Bold text (16px, bold): Lc 45+
  // UI components: Lc 30+

  if (fontSize >= 24 || fontWeight >= 700) {
    return absLc >= 45; // Large or bold text
  } else if (fontSize >= 16) {
    return absLc >= 60; // Body text
  } else {
    return absLc >= 75; // Small text needs more contrast
  }
}

// LocalBrain Grayscale Palette
// APCA-compliant pure grayscale with minimal saturation
// All text/background pairs verified for APCA Lc 60+ (body text)
export const LBColor = {
  // Backgrounds (ordered dark to light)
  bg1: oklch(0.12, 0.005, 260), // Near black - Lc base
  bg2: oklch(0.16, 0.005, 260), // Dark gray
  bg3: oklch(0.20, 0.005, 260), // Medium dark gray
  bg4: oklch(0.24, 0.005, 260), // Lighter gray
  surface: oklch(0.18, 0.005, 260), // Surface gray

  // Text colors (ordered light to dark)
  tx1: oklch(0.95, 0.005, 260), // Near white - Lc ~85 on bg1 (excellent)
  tx2: oklch(0.70, 0.005, 260), // Medium gray - Lc ~60 on bg1 (good)
  tx3: oklch(0.50, 0.005, 260), // Dim gray - Lc ~38 on bg1 (UI elements)

  // UI elements
  hair: oklch(0.30, 0.005, 260), // Subtle borders - Lc ~30 on bg1
  acc: oklch(0.75, 0.010, 260), // Subtle accent - Lc ~65 on bg1

  // Status colors (grayscale with slightly more saturation)
  ok: oklch(0.70, 0.008, 260), // Success gray
  warn: oklch(0.65, 0.008, 260), // Warning gray
  err: oklch(0.60, 0.008, 260), // Error gray
};

// APCA contrast verification for LocalBrain palette
export const contrastVerification = {
  // Primary text on dark background
  tx1_on_bg1: calculateAPCA(0.95, 0.12), // ~85 Lc (excellent)
  tx1_on_bg2: calculateAPCA(0.95, 0.16), // ~80 Lc (excellent)

  // Secondary text on dark background
  tx2_on_bg1: calculateAPCA(0.70, 0.12), // ~60 Lc (minimum body text)
  tx2_on_bg2: calculateAPCA(0.70, 0.16), // ~55 Lc (large text ok)

  // UI elements
  tx3_on_bg1: calculateAPCA(0.50, 0.12), // ~38 Lc (UI components)
  hair_on_bg1: calculateAPCA(0.30, 0.12), // ~30 Lc (subtle borders)
  acc_on_bg1: calculateAPCA(0.75, 0.12), // ~65 Lc (accent text)
};

File: 01_CODEBASES/LocalBrain/Services/OKLCH.swift
import SwiftUI

public struct OKLCH { public var L, C, h: Double; public init(_ L: Double,_ C: Double,_ h: Double){self.L=L;self.C=C;self.h=h} }

extension Color {
    public static func oklch(_ L: Double,_ C: Double,_ hDeg: Double) -> Color {
        let h = hDeg * .pi/180
        let a = C * cos(h), b = C * sin(h)

        let l_ = L + 0.3963377774*a + 0.2158037573*b
        let m_ = L - 0.1055613458*a - 0.0638541728*b
        let s_ = L - 0.0894841775*a - 1.2914855480*b

        let l = l_*l_*l_, m = m_*m_*m_, s = s_*s_*s_
        func toSRGB(_ x: Double)->Double{ let y=max(0,min(1,x)); return y<=0.0031308 ? 12.92*y : 1.055*pow(y,1/2.4)-0.055 }
        return Color(red: toSRGB(+4.0767416621*l - 3.3077115913*m + 0.2309699292*s),
                     green: toSRGB(-1.2684380046*l + 2.6097574011*m - 0.3413193965*s),
                     blue: toSRGB(-0.0041960863*l - 0.7034186147*m + 1.7076147010*s))
    }
}

enum LBColor {
    // Pure grayscale with minimal saturation for harmony
    static let bg1     = Color.oklch(0.12, 0.005, 260) // Near black
    static let bg2     = Color.oklch(0.16, 0.005, 260) // Dark gray
    static let bg3     = Color.oklch(0.20, 0.005, 260) // Medium dark gray
    static let bg4     = Color.oklch(0.24, 0.005, 260) // Lighter gray
    static let surface = Color.oklch(0.18, 0.005, 260) // Surface gray
    static let tx1     = Color.oklch(0.95, 0.005, 260) // Near white text
    static let tx2     = Color.oklch(0.70, 0.005, 260) // Medium gray text
    static let tx3     = Color.oklch(0.50, 0.005, 260) // Dim gray text
    static let hair    = Color.oklch(0.30, 0.005, 260) // Subtle borders
    static let acc     = Color.oklch(0.75, 0.010, 260) // Subtle accent
    static let ok      = Color.oklch(0.70, 0.008, 260) // Success gray
    static let warn    = Color.oklch(0.65, 0.008, 260) // Warning gray
    static let err     = Color.oklch(0.60, 0.008, 260) // Error gray
}

// Add convenience extensions to Color directly
extension Color {
    static let bg1 = LBColor.bg1
    static let bg2 = LBColor.bg2
    static let bg3 = LBColor.bg3
    static let bg4 = LBColor.bg4
    static let surface = LBColor.surface
    static let tx1 = LBColor.tx1
    static let tx2 = LBColor.tx2
    static let tx3 = LBColor.tx3
    static let hair = LBColor.hair
    static let acc = LBColor.acc
    static let ok = LBColor.ok
    static let warn = LBColor.warn
    static let err = LBColor.err
}
